### ISSUE TITLE
Confuse about the train function of QAT_torch_quantizer.py 

### ISSUE THREAD

**Issue by Lycan1003** [microsoft/nni#3071]
Excuse me, I am confusing about nni/examples/model_compress/QAT_torch_quantizer.py **line 92**
`train(model, quantizer, device, train_loader, optimizer)`
It is using "quantizer" as the parameter of the train function, 
but actually, the train function definition (**line 29**) does not use the "quantizer" to do anything.
```
def train(model, quantizer, device, train_loader, optimizer):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 100 == 0:
            print('{:2.0f}%  Loss {}'.format(100 * batch_idx / len(train_loader), loss.item()))
```
so I am wondering **Which step is the QAT operation performed?**
I'm a rookie of PyTorch, please give me some advice, thanks a lot.

**Comment by linbinskn** [comment]
When you init the QAT by using **line86**
`quantizer = QAT_Quantizer(model, configure_list, optimizer)`, all the modules in your model will be converted to quantized module. After converting, during the training process, the process of quantization will be done by the quantized module. So the quantizer parameter doesn't needed in the function **train**. If you want to know the details of quantization in NNI, I recommend you to read the code in '/nni/src/sdk/pynni/nni/compression/torch/compressor.py' and '/nni/src/sdk/pynni/nni/compression/torch/quantization/quantizers.py'.

**Comment by Lycan1003** [comment]
> When you init the QAT by using **line86**
> `quantizer = QAT_Quantizer(model, configure_list, optimizer)`, all the modules in your model will be converted to quantized module. After converting, the process of quantization will be done by the quantized module during the training process. So the quantizer parameter isn't needed in the function **train**. If you want to know the details of quantization in NNI, I recommend you to read the code in '/nni/src/sdk/pynni/nni/compression/torch/compressor.py' and '/nni/src/sdk/pynni/nni/compression/torch/quantization/quantizers.py'.

Thanks for your reply! I think I get the key. 
And I find that the final model after training is still the float model, so **how to save the quantized model with int8?**

**Comment by linbinskn** [comment]
So far, NNI only supports simulation quantization. I recommend you using it to verity the quantization scheme and observing the trend of accuracy. You can save the model whose inference result is the same with quantized model during inference, **but it will still be float32 type and the model size will not change**. NNI plans to support the real quantization in the near future which support actual model speed up and reducing model size in the following releases. May be you can pay attention to the update if you are interested.

**Comment by Lycan1003** [comment]
> So far, NNI only supports simulation quantization. I recommend you using it to verity the quantization scheme and observing the trend of accuracy. You can save the model whose inference result is the same with quantized model during inference, **but it will still be float32 type and the model size will not change**. NNI plans to support the real quantization in the near future which supports actual model speed up and reducing model size in the following releases. May be you can pay attention to the update if you are interested.

Thanks for your answers, I will look forward to follow-up updates.
Have a nice day!
